app/admin/tenants/onboarding.py

```diff
@@ -0,0 +1,203 @@
+
+            # Remove duplicates from the role list
+            updated_actor["roles"] = list(set(update.roles))
+            updated_actor["did"] = onboard_result.did
+            updated_actor["didcomm_invitation"] = onboard_result.didcomm_invitation
+
+        await update_actor(updated_actor)
+
+
+async def onboard_ecosystem_tenant(
+    *, name: str, roles: List[TrustRegistryRole], tenant_auth_token: str
+) -> OnboardResult:
+    if "issuer" in roles:
+        # Get yoma and tenant controllers, onboard issuer
+        async with get_yoma_controller() as yoma_controller, get_tenant_controller(
+            Role.ECOSYSTEM, tenant_auth_token
+        ) as tenant_controller:
+            return await onboard_issuer(
+                name=name,
+                issuer_controller=tenant_controller,
+                endorser_controller=yoma_controller,
+            )
+
+    elif "verifier" in roles:
+        async with get_tenant_controller(
+            Role.ECOSYSTEM, tenant_auth_token
+        ) as tenant_controller:
+            return await onboard_verifier(
+                name=name, verifier_controller=tenant_controller
+            )
+
+    raise CloudApiException("Unable to onboard tenant without role(s).")
+
+
+async def onboard_issuer(
+    *, name: str, issuer_controller: AcaPyClient, endorser_controller: AcaPyClient
+):
+    """Onboard the controller as issuer.
+
+    The onboarding will take care of the following:
+      - make sure the issuer has a public did
+      - make sure the issuer has a connection with the endorser
+      - make sure the issuer has set up endorsement with the endorser connection
+
+    Args:
+        name (str): name of the issuer
+        issuer_controller (AcaPyClient): authenticated ACA-Py client for issuer
+        endorser_controller (AcaPyClient): authenticated ACA-Py client for endorser
+    """
+    # Make sure the issuer has a public did
+    try:
+        issuer_did = await acapy_wallet.get_public_did(controller=issuer_controller)
+    except CloudApiException:
+        # no public did
+        issuer_did = await acapy_wallet.create_did(issuer_controller)
+        await acapy_ledger.register_nym_on_ledger(
+            endorser_controller,
+            did=issuer_did.did,
+            verkey=issuer_did.verkey,
+            alias=name,
+        )
+        await acapy_ledger.accept_taa_if_required(issuer_controller)
+        await acapy_wallet.set_public_did(issuer_controller, issuer_did.did)
+
+    endorser_did = await acapy_wallet.get_public_did(controller=endorser_controller)
+
+    # Make sure the issuer has a connection with the endorser
+    invitation = await endorser_controller.out_of_band.create_invitation(
+        auto_accept=True,
+        body=InvitationCreateRequest(
+            handshake_protocols=["https://didcomm.org/didexchange/1.0"],
+            use_public_did=True,
+        ),
+    )
+
+    # FIXME: Wait for the connection to be completed here once the webhooks PR is merged
```

Comment:
Is there already a way to integrate with webhooks outside of the tests? Would be good to add some nice util methods for this to the webhook facade. I would like to do this as part of this issue: https://github.com/didx-xyz/aries-cloudapi-python/issues/159

Reply:
yes you can pass to subroute `/webhooks` your controller and get your associated hooks. see `app/generic/webhooks.py`

Reply:
Yeah but I meant more something like the `check_webhook_state` test util. Looping through a list of webhooks is quite clunky and can be easily generalized so it can be reused everywhere.

My biggest concern with the approach you mention is that it fetches all webhooks multiple times. I would rather a subscriber model like the following:

1. The cloud api listens to all webhooks constantly and doesn't store them
2. methods in the cloudapi can start listening to webhooks (subscribe) and have a callback method called each time

For waiting for a connection that means:

1. Create connection invitation
2. Start listening to connection webhook events for a wallet
3. Cloud api publishes all connections events to the callback I provided
4. Share invitation with other party
5. wait for the connection state to be completed

This has a few advantages IMO:
1. The cloudapi doesn't store webhook events by default
2. If a service needs to listen to webhooks it explicitly subscribes to a topic / wallet id (like an event listener) and also explicitly unsubscribes. This means you only store webhook events for a short period and release them immediately afterwards. It also means all other webhook events (that aren't relevant atm) are immediately flushed from memory.

Hope this makes sense.

Reply:
yeah makes sense